### DISCUSSION TITLE
Add configuration option for JSON data format in otlphttp client

### DISCUSSION THREAD

**Discussion by gmreads** [open-telemetry/opentelemetry-go#3969] [Ideas]
Hi,

As per discussions on slack with @Aneurysm9 , having JSON support for exporter in addition to protobuff is welcome.

I came up with this rough draft before actually working on code.
 @Aneurysm9 Would appreciate your thoughts on this, is this the right direction?

**Requirements**
- Backwards compatible by setting defaults as protobuff
- Single config to set Content-Type header and send JSON to the receiver.

**Proposed Interface**
- Add a new option while configuring client 
   ```golang
    client := otlptracehttp.NewClient(
    // default being protobuff, and the supported data formats somewhere in constants
        otlptracehttp.WithDataFormat("JSON")
   ```

**Implementation Details**
- Planning to make changes in [uploadTraces](https://github.com/open-telemetry/opentelemetry-go/blob/a25f4fe7efddc017e1b557ba94088159c6da4174/exporters/otlp/otlptrace/otlptracehttp/client.go#L133) function here
  ```golang
   // UploadTraces sends a batch of spans to the collector.
   func (d *client) UploadTraces(ctx context.Context, protoSpans []*tracepb.ResourceSpans) error {
	pbRequest := &coltracepb.ExportTraceServiceRequest{
		ResourceSpans: protoSpans,
	}
        // Marshal to JSON based on d.cfg.dataFormat
	rawRequest, err := proto.Marshal(pbRequest) 
	if err != nil {
		return err
	}
  ```

**Unknowns**
- Do proto files have proper struct tags for json ?
- What's the difference between proto3 and proto
- Why two different json tags are present ? ( `json:"name=resource_spans"` `json:"name=resourceSpans"`)
(in the `ExportTraceServiceRequest` struct

**Comment by pellared** [comment]
I guess for doing JSON serialization you should use https://pkg.go.dev/google.golang.org/protobuf/encoding/protojson

You can also look at https://github.com/open-telemetry/opentelemetry-collector/blob/main/pdata/ptrace/internal/ptracejson/json.go

I think it is also worth to mention the specification: https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/protocol/otlp.md#json-protobuf-encoding

**Reply by gmreads** [reply]
Yes I was planning to use the protojson package.

Thanks for the references, but they are primarily implementation concerns. 

I wanted to take inputs on whether the proposed interface is acceptable to be merged upstream ?

**Reply by pellared** [reply]
In general looks fine. Maybe with `WithEncoding`  instead of `WithDataFormat`. I think it would be good if it accepts something more or less like

```go
type Encoding string

const (
   EncodingProtobuf Encoding = "protobuf"
   EncodingJson  = "json"
```

**Reply by gmreads** [reply]
Hey I'll be getting started on this soon.

In the meantime, since we already have our own receiver(nodeJS) was trying to do the conversion to JSON there.

Trying to parse the protobuff 
`Getting this error: Error: invalid wire type 7 at offset 238`

proto files used : https://github.com/open-telemetry/opentelemetry-proto/blob/main/opentelemetry/proto/collector/trace/v1/trace_service.proto

**Reply by gmreads** [reply]
Ignore, I was reading the data normally. Whereas nodejs http.Request.Body is a stream so
- use Buffer.concat instead of Buffer.from to read the request body.
